test_data/asm/keccak.asm

```diff
@@ -0,0 +1,200 @@
+use std::array;
+use std::utils;
+
+// function definition
+let or: int, int -> int = |a, b| a | b;
+let and: int, int -> int = |a, b| a & b;
+let xor: int, int -> int = |a, b| a ^ b;
+let not: int -> int = |a| a ^ 0xffffffffffffffff; // bitwise not for 64 bits
+
+// ln 9 - 12
+// TODO: check that these are good for word size 32
+let RHO: int[] = [
+    1,  3,  6,  10, 15, 21, 28, 36, 45, 55, 2,  14,
+    27, 41, 56, 8,  25, 43, 62, 18, 39, 61, 20, 44
+];
+
+// ln 14 - 16
+// TODO: check that these are good for word size 32
+let PI: int[] = [
+    10, 7,  11, 17, 18, 3, 5,  16, 8,  21, 24, 4,
+    15, 23, 19, 13, 12, 2, 20, 14, 22, 9,  6,  1
+];
+
+// inverse of PI using 1 index plus zero at the beginning
+// because rho pi step doesn't update st[0]
+let PI_INVERSE: int[] = [
+    0, 
+    24, 18, 6, 12, 7, 23, 2, 9, 22, 1, 3, 17, 
+    16, 20, 13, 8, 4, 5, 15, 19, 10, 21, 14, 11
+];
+
+// ln 19 - 28
+// TODO: check that these are good for word size 32
+let RC: int[] = [
+    0x0000000000000001, 0x0000000000008082, 0x800000000000808a,
+    0x8000000080008000, 0x000000000000808b, 0x0000000080000001,
+    0x8000000080008081, 0x8000000000008009, 0x000000000000008a,
+    0x0000000000000088, 0x0000000080008009, 0x000000008000000a,
+    0x000000008000808b, 0x800000000000008b, 0x8000000000008089,
+    0x8000000000008003, 0x8000000000008002, 0x8000000000000080,
+    0x000000000000800a, 0x800000008000000a, 0x8000000080008081,
+    0x8000000000008080, 0x0000000080000001, 0x8000000080008008
+];
+
+// ln 30 - 33
+// left rotation
+// 32 and 0xffffffff if u32
+let rotl64: int, int -> int = |x, n| and(or((x << n), (x >> (64 - n))), 0xffffffffffffffff); 
```

Comment:
Shouldn't this move to the beginning of the file? Because I think this is one of the elementary gates.

Reply:
Agree